tests/Analyze/Gen.hs

```diff
@@ -0,0 +1,464 @@
+       let (size1, size2) = arithSize op size
+       Gen.subtermM2 (genCore (BoundedDecimal size1)) (genCore (BoundedInt size2)) $
+         \x y -> mkDec $ DecIntArithOp op (extract x) (extract y)
+  , do
+       op <- genArithOp
+       let (size1, size2) = arithSize op size
+       Gen.subtermM2 (genCore (BoundedInt size1)) (genCore (BoundedDecimal size2)) $
+         \x y -> mkDec $ IntDecArithOp op (extract x) (extract y)
+  , Gen.subtermM2 (genCore bounded) (genCore (BoundedInt (0 +/- 255))) $ \x y -> do
+      op <- genRoundingLikeOp
+      mkDec $ RoundingLikeOp2 op (extract x) (extract y)
+  ]
+genCore (BoundedString len) = Gen.recursive Gen.choice [
+    ESimple TStr . CoreTerm . Lit <$> Gen.string (Range.exponential 1 len) Gen.unicode
+  ] [
+    scale 4 $ Gen.subtermM2
+      (genCore (BoundedString (len `div` 2)))
+      (genCore (BoundedString (len `div` 2))) $ \x y ->
+        pure $ ESimple TStr $ Inj $ StrConcat (extract x) (extract y)
+  ]
+genCore BoundedBool = Gen.recursive Gen.choice [
+    ESimple TBool . CoreTerm . Lit <$> Gen.bool
+  ] $ scale 4 <$> [
+    do op <- genComparisonOp
+       Gen.subtermM2 (genCore intSize) (genCore intSize) $ \x y -> do
+         mkBool $ IntegerComparison op (extract x) (extract y)
+  , do op <- genComparisonOp
+       Gen.subtermM2 (genCore decSize) (genCore decSize) $ \x y ->
+         mkBool $ DecimalComparison op (extract x) (extract y)
+  , do op <- genComparisonOp
+       Gen.subtermM2 (genCore BoundedTime) (genCore BoundedTime) $ \x y ->
+         mkBool $ TimeComparison op (extract x) (extract y)
+  , do op <- genComparisonOp
+       Gen.subtermM2 (genCore strSize) (genCore strSize) $ \x y ->
+         mkBool $ StringComparison op (extract x) (extract y)
+  , do op <- Gen.element [Eq, Neq]
+       Gen.subtermM2 (genCore BoundedBool) (genCore BoundedBool) $ \x y ->
+         mkBool $ BoolComparison op (extract x) (extract y)
+  , do op <- Gen.element [AndOp, OrOp]
+       Gen.subtermM2 (genCore BoundedBool) (genCore BoundedBool) $ \x y ->
+         mkBool $ Logical op [extract x, extract y]
+  , Gen.subtermM (genCore BoundedBool) $ \x ->
+      mkBool $ Logical NotOp [extract x]
+  ]
+genCore BoundedTime = Gen.recursive Gen.choice [
+    ESimple TTime . CoreTerm . Lit <$> Gen.enumBounded -- Gen.int64
+  ] $ scale 4 <$> [
+    Gen.subtermM2 (genCore BoundedTime) (genCore (BoundedInt 1e9)) $ \x y ->
+      pure $ ESimple TTime $ Inj $ IntAddTime (extract x) (extract y)
+  , Gen.subtermM2 (genCore BoundedTime) (genCore (BoundedDecimal 1e9)) $ \x y ->
+      pure $ ESimple TTime $ Inj $ DecAddTime (extract x) (extract y)
+  ]
+genCore BoundedKeySet = ESimple TKeySet . CoreTerm . Lit . KeySet
+  <$> genInteger (0 ... 2)
+
+intSize, decSize, strSize :: BoundedType
+intSize = BoundedInt     (0 +/- 1e25)
+decSize = BoundedDecimal (0 +/- 1e25)
+strSize = BoundedString  1000
+
+-- TODO
+-- generic:
+-- # Let
+-- object:
+-- # Read
+-- string:
+-- # Write
+-- time:
+-- # ParseTime
```

Comment:
in these TODOs, any blockers to merging?

Reply:
No -- these are just reminders for myself of things we don't cover yet.